Legacy/bonej/src/main/java/org/bonej/plugins/ParticleCounter.java

```diff
@@ -410,48 +441,337 @@ else if (choice.equals(items2[1]))
+		final double l2 = lengths[1];
+		final double l3 = lengths[2];
+
+		final List<Point3f> mesh = new ArrayList<>();
+		final Point3f start1 = new Point3f();
+		start1.x = (float) (cX - eVec1x * l1);
+		start1.y = (float) (cY - eVec1y * l1);
+		start1.z = (float) (cZ - eVec1z * l1);
+		mesh.add(start1);
+
+		final Point3f end1 = new Point3f();
+		end1.x = (float) (cX + eVec1x * l1);
+		end1.y = (float) (cY + eVec1y * l1);
+		end1.z = (float) (cZ + eVec1z * l1);
+		mesh.add(end1);
+
+		final Point3f start2 = new Point3f();
+		start2.x = (float) (cX - eVec2x * l2);
+		start2.y = (float) (cY - eVec2y * l2);
+		start2.z = (float) (cZ - eVec2z * l2);
+		mesh.add(start2);
+
+		final Point3f end2 = new Point3f();
+		end2.x = (float) (cX + eVec2x * l2);
+		end2.y = (float) (cY + eVec2y * l2);
+		end2.z = (float) (cZ + eVec2z * l2);
+		mesh.add(end2);
+
+		final Point3f start3 = new Point3f();
+		start3.x = (float) (cX - eVec3x * l3);
+		start3.y = (float) (cY - eVec3y * l3);
+		start3.z = (float) (cZ - eVec3z * l3);
+		mesh.add(start3);
+
+		final Point3f end3 = new Point3f();
+		end3.x = (float) (cX + eVec3x * l3);
+		end3.y = (float) (cY + eVec3y * l3);
+		end3.z = (float) (cZ + eVec3z * l3);
+		mesh.add(end3);
+
+		final Color3f aColour = new Color3f(1.0f, green, 0.0f);
+		try {
+			univ.addLineMesh(mesh, aColour, title, false).setLocked(true);
+		}
+		catch (final NullPointerException npe) {
+			IJ.log("3D Viewer was closed before rendering completed.");
+		}
+	}
+
+	/**
+	 * Draw the particle centroids in a 3D viewer
+	 *
+	 * @param centroids [n][3] centroids of particles.
+	 * @param univ universe where the centroids are displayed.
+	 */
+	private static void displayCentroids(final double[][] centroids,
+		final Image3DUniverse univ)
+	{
+		final int nCentroids = centroids.length;
+		for (int p = 1; p < nCentroids; p++) {
+			IJ.showStatus("Rendering centroids...");
+			IJ.showProgress(p, nCentroids);
+			final Point3f centroid = new Point3f();
+			centroid.x = (float) centroids[p][0];
+			centroid.y = (float) centroids[p][1];
+			centroid.z = (float) centroids[p][2];
+			final List<Point3f> point = new ArrayList<>();
+			point.add(centroid);
+			final CustomPointMesh mesh = new CustomPointMesh(point);
+			mesh.setPointSize(5.0f);
+			final float red = 0.0f;
+			final float green = 0.5f * p / nCentroids;
+			final float blue = 1.0f;
+			final Color3f cColour = new Color3f(red, green, blue);
+			mesh.setColor(cColour);
+			try {
+				univ.addCustomMesh(mesh, "Centroid " + p).setLocked(true);
+			}
+			catch (final NullPointerException npe) {
+				IJ.log("3D Viewer was closed before rendering completed.");
+				return;
+			}
+		}
+	}
+
+	private static void displayEllipsoids(final Object[][] ellipsoids,
+		final Image3DUniverse univ)
+	{
 		final int nEllipsoids = ellipsoids.length;
-		ellipsoidLoop: for (int el = 1; el < nEllipsoids; el++) {
+		for (int el = 1; el < nEllipsoids; el++) {
 			IJ.showStatus("Rendering ellipsoids...");
```

Comment:
why does this loop start at 1 ?????

Reply:
I don't know exactly, but it seems to consistent in other parts of the code where the `ellipsoid[][]` is looped. Maybe `ellipsoid[0]` is the background particle...